File: 2024/day7/day7.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

type Equation struct {
	TestValue int
	Numbers   []int
}

func main() {
	equations := readInput()

	add := func(x, y int) int { return x + y }
	multiply := func(x, y int) int { return x * y }
	concatenate := func(x, y int) int {
		combined, _ := strconv.Atoi(fmt.Sprintf("%d%d", x, y))
		return combined
	}

	part1, part2 := 0, 0
	for _, eq := range equations {
		if canBeMadeTrue(eq, []func(int, int) int{add, multiply}) {
			part1 += eq.TestValue
		}
		if canBeMadeTrue(eq, []func(int, int) int{add, multiply, concatenate}) {
			part2 += eq.TestValue
		}
	}

	fmt.Printf("Part 1: %d\n", part1)
	fmt.Printf("Part 2: %d\n", part2)
}

func canBeMadeTrue(eq Equation, operations []func(int, int) int) bool {
	var solve func(result int, numbers []int) bool

	solve = func(result int, numbers []int) bool {
		if len(numbers) == 0 {
			return result == eq.TestValue
		}
		if result > eq.TestValue {
			return false
		}
		for _, operation := range operations {
			if solve(operation(result, numbers[0]), numbers[1:]) {
				return true
			}
		}
		return false
	}

	return solve(eq.Numbers[0], eq.Numbers[1:])
}

func readInput() []Equation {
	scanner := bufio.NewScanner(os.Stdin)
	var equations []Equation

	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		parts := strings.Split(line, ": ")
		testValue, _ := strconv.Atoi(parts[0])
		numberStrings := strings.Split(parts[1], " ")
		numbers := make([]int, len(numberStrings))
		for i, numStr := range numberStrings {
			numbers[i], _ = strconv.Atoi(numStr)
		}
		equations = append(equations, Equation{TestValue: testValue, Numbers: numbers})
	}

	return equations
}

File: 2024/day7/day7.py
import math
import sys

equations = []
for line in sys.stdin:
    test_value, numbers = line.strip().split(': ')
    equations.append((int(test_value), list(map(int, numbers.split(' ')))))

def can_be_made_true(test_value, numbers, operations):
    def solve(result, numbers):
        if len(numbers) == 0:
            return result == test_value
        if result > test_value:
            return False
        for op in operations:
            if solve(op(result, numbers[0]), numbers[1:]):
                return True
        return False

    return solve(numbers[0], numbers[1:])

add = lambda x, y: x + y
multiply = lambda x, y: x * y
concatenate = lambda x, y: x * (10 ** (math.floor(math.log10(y)) + 1)) + y
# concatenate = lambda x, y: int(f'{x}{y}')

print(f'Part 1: {sum(t for t, ns in equations if can_be_made_true(t, ns, [add, multiply]))}')
print(f'Part 2: {sum(t for t, ns in equations if can_be_made_true(t, ns, [add, multiply, concatenate]))}')
